core/src/main/java/edu/wpi/grip/core/operations/composite/PublishVideoOperation.java

```diff
@@ -0,0 +1,157 @@
+import com.google.common.eventbus.Subscribe;
+import edu.wpi.grip.core.InputSocket;
+import edu.wpi.grip.core.Operation;
+import edu.wpi.grip.core.OutputSocket;
+import edu.wpi.grip.core.SocketHints;
+import edu.wpi.grip.core.events.StepRemovedEvent;
+import org.bytedeco.javacv.FrameConverter;
+import org.bytedeco.javacv.Java2DFrameConverter;
+import org.bytedeco.javacv.OpenCVFrameConverter;
+
+import javax.imageio.ImageIO;
+import java.awt.image.BufferedImage;
+import java.awt.image.RenderedImage;
+import java.io.ByteArrayOutputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.Optional;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+import static org.bytedeco.javacpp.opencv_core.Mat;
+
+/**
+ * Publish an M-JPEG stream with the protocol used by SmartDashboard and the FRC Dashboard.  This allows FRC teams to
+ * view video streams on their dashboard during competition even when GRIP has exclusive access to the camera.  In
+ * addition, an intermediate processed image in the pipeline could be published instead.
+ * <p>
+ * Based on WPILib's CameraServer class: https://github.com/robotpy/allwpilib/blob/master/wpilibj/src/athena/java/edu/wpi/first/wpilibj/CameraServer.java
+ */
+public class PublishVideoOperation implements Operation {
+
+    private final Logger logger = Logger.getLogger(PublishVideoOperation.class.getName());
+
+    private static final int PORT = 1180;
+    private static final byte[] MAGIC_NUMBER = {0x01, 0x00, 0x00, 0x00};
+
+    private final Object imageLock = new Object();
+    private RenderedImage image = null;
+    private Optional<Thread> serverThread = Optional.empty();
+
+    /**
+     * Listens for incoming connections on port 1180 and writes JPEG data whenever there's a new frame.
+     */
+    private final Runnable runServer = () -> {
+        // Loop forever (or at least until the thread is interrupted).  This lets us recover from the dashboard
+        // disconnecting or the network connection going away temporarily.
+        while (true) {
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            ServerSocket serverSocket = null;
+            Socket socket = null;
+
+            try {
+                serverSocket = new ServerSocket(PORT);
+//                serverSocket.setReuseAddress(true);
+//                serverSocket.bind(new InetSocketAddress(PORT));
+                logger.info("Starting camera server");
+
+                while (!serverSocket.isClosed() && !Thread.interrupted()) {
+                    socket = serverSocket.accept();
+                    DataOutputStream socketStream = new DataOutputStream(socket.getOutputStream());
+                    logger.info("Got connection from " + socket.getInetAddress());
+
+                    while (!socket.isClosed() && !Thread.interrupted()) {
```

Comment:
`Thread.interrupted()` resets the flag to false.
You need to use `!Thread.currentThread().isInterrupted()` so the flag isn't reset.
Otherwise the outer while loop will not be exited.